### ISSUE TITLE
Bring back 'About' tab and reuse widget space

### ISSUE THREAD

**Issue by ghost** [kamilion/customizer#125]
This is another follow up from issue #121 regarding translatable 'About'.

Following issue #124 and #128 would be this one, which is bringing back 'About' tab.

**Proposed implementation**

Reusing widget space is the approach used by modern applications based on GTK+ 3. This way, no pop-up window and no extra tabs are needed. Similar GUI behaviour can be reproduced in PyQt 4 or 5.

Use [QStackedWidget](http://doc.qt.io/qt-4.8/qstackedwidget.html) to handle multiple pages, so that the same widget area can be reused to display different pages, one at a time. The page display can be changed by clicking on buttons, which are required for pairing the signals and slots.

**Expected appearance**

The following sketchy UI illustrates how Customizer GUI should look like.

![gui-420-sketchy-all](https://cloud.githubusercontent.com/assets/1223304/20393280/3c5bf9cc-ad16-11e6-9912-077249189c91.png)

Like other existing tabs, 'About' tab will use grid layout to ensure the positioning of widget items are distributed appropriately when Customizer GUI is maximized to fill the screen.

There will be other minor things to be implemented together, which cannot be illustrated in sketchy UI. For any following changes and explanations, refer to relevant comments at below.

**This is expected to be implemented in patch release 4.1.5. Widget now prefers horizontal arrangement (see comment below).**

**Comment by clearkimura** [comment]
In the first screenshot of "**Details**" page, the **Source** field is subject to changes.

The field may or may not be created, depending on how the forked repository will name the version to differentiate between 'old stable' and 'forked' releases. If the forked releases adapt something similar to Ubuntu package naming i.e. `hello-1.0-ubuntu1` compared to Debian source package i.e. `hello-1.0`, then the **Source** field is not necessary.

**Comment by ghost** [comment]
**Tested early implementation**

For handling multiple pages, QStackedWidgets is used to contain those pages and several QPushButton are used to change current view to respective pages:

```
self.pushButton_1.clicked.connect(self.GotoPage1)
self.pushButton_2.clicked.connect(self.GotoPage2)

def GotoPage1(self):
    self.stackedWidget.setCurrentIndex(0)

def GotoPage2(self):
    self.stackedWidget.setCurrentIndex(1)
```

For displaying external file _without_ Qt Resouce system, QTextBrowser is used like this:

```
self.textBrowser.setSource(QtCore.QUrl("/usr/share/doc/customizer/copyright"))
```

Above are showing the critical lines of code, which I had tested in separate Qt Designer file (.ui) and Python script (.py). As such, there will be few differences here and there when bringing this idea into existing source code.

For example, in actual implementation, the QUrl should be set to `@PREFIX@` path, which is the easier way to implement (_without_ Qt Resouce system). This is the similar approach used by the original developer for setting the Window and Alt-Tab icon in `gui.py.in` source file:

```
MainWindow.setWindowIcon(QtGui.QIcon("@PREFIX@/share/customizer/logo.png"))
```

With early implementation has been successfully done and understood by me, I am now more confident to implement the changes to UI in coming weeks. Patch relese 4.1.4 shall be well prepared for making these changes to UI. Glad to know that I am making some progress.

**Comment by ghost** [comment]
**Minor things explained**

The following are some minor things that I have realized, after doing repeated testing of early implementation.

* Don't use [QTextBrowser](http://doc.qt.io/qt-4.8/qtextbrowser.html) for displaying text file

  It was easy to specify the source file to be read by QTextBrowser. But there is one persistent issue by using this widget: It always ignores the newlines in plain text. So multiple lines of text will be rendered as a single, long line of text in QTextBrowser. And no, we shall never use HTML file. Period.

* Use [QPlainTextEdit](http://doc.qt.io/qt-4.8/qplaintextedit.html) instead

  Besides the unwanted behaviour of ignoring newlines, QTextBrowser has many features that are beyond simple use of displaying text file, such as hyperlinks, navigation or even reload the text. We don't need such features at all, so it is better to use something... less sophisticated.

  By using QPlainTextEdit, it is possible to read the file and display as plain text in one line:

  ```
  self.plainTextEdit.setPlainText(open('FILE').read())
  ```

  In which, FILE can be just target file name i.e. `file.txt` that exists within same directory as Python script, or it can be absolute path to target file i.e. `/usr/share/customizer/file.txt`. I have successfully tested both methods by using separate script and UI file.

* Consider to have exception when target file is not found

  When target file specified in QPlainTextEdit is not found, the Qt program will throw an error and will refuse to run the program at all. To prevent this from happening, there should be additional lines to check if the target file exist or not, then display "File not found" in the widget if the file doesn't exist.

  So instead of one-liner code shown above, a better approach will be like this:

  ```
  if os.path.isfile('FILE'):
      self.plainTextEdit.setPlainText(open('FILE').read())
  else:
      self.plainTextEdit.setPlainText("File not found")
  ```

  Note that `import os.path` is required to use [`os.path.isfile`](https://docs.python.org/2/library/os.path.html#os.path.isfile).

* Consider to define function to reuse code

  This is optional. If this is successfully implemented, we would not have to write duplicate lines of code; Just define one block of function that can be called again to read file for two instances i.e. 'Credits' and 'License' tabs.

  The successful implementation I have tried:

  ```
  self.ShowFile('FILE')

  def ShowFile(self, args):
      if os.path.isfile(args):
          self.plainTextEdit.setPlainText(open(args).read())
      else:
          self.plainTextEdit.setPlainText("File not found")
  ```

  **Update 2** I'd have to add second argument so that the function is reusable for multiple widget of same kind i.e. plainTextEdit_1, plainTextEdit_2, etc. The modifed implementation I have tried:

  ```
  self.ShowFile(self.plainTextEdit_1, 'FILE_1')
  self.ShowFile(self.plainTextEdit_2, 'FILE_2')

  def ShowFile(self, obj, args):
      if os.path.isfile(args):
          obj.setPlainText(open(args).read())
      else:
          obj.setPlainText("File not found")
  ```

  The modified implementation above was found after doing trials and errors by myself; I have not found any relevant example that creates custom widget as such. Most examples are using loop with counter to clone the widgets instances and *not* used for displaying files like above.

* Use `autoDefault` for [QPushButton](http://doc.qt.io/qt-4.8/qpushbutton.html)

  When navigating UI using Tab key or mouse clicks, the selected widget should be highlighted to indicate it has been selected. But default behaviour of Qt doesn't reflect this. Following this, I have discovered that `autoDefault` option will enable the intended behaviour.

  I am considering to apply this intended behaviour for other buttons in existing tabs, because I failed to see any highlight or focus when navigating with Tab key or mouse clicks. The tab widget itself will show the tab is being selected, but not visible for other widgets.

**TL;DR**

* QPlainTextEdit shall be used instead of QTextBrowser
* Exception of "File not found" shall be considered
* If possible, define a function to allow reusing code (Update 1: Now successful)
* Turn on `autoDefault` for QPushButton in all tabs, not just 'About' tab

**Update 1 (2016.11.19)**

I managed to figure out how to implement as function; I was missing a second argument to be passed to the function. With this, I think all requirements to solve this issue have been tested and could be implemented successfully when working on 4.1.5 later. Glad to see I made another progress.

**Update 2 (2016.11.23)**

I found a limitation with existing function, which allows only one instance per widget type. I had to pass second argument to the function so that multiple instances of same widget type can be set.

**Comment by ghost** [comment]
**Reusing code explained**

The concept is applicable to other block of codes that contain duplicates, such as showing content of text file and set the content in the QPlainTextEdit widget. In this example, only public slot is involved.

When bringing this concept "as it is" to paired signals and slots, such as switching display of stacked widget by clicking respective buttons, I will be shown this error:

```
  File "demo.py", line 20, in __init__
    self.pushButton_1.clicked.connect(self.GotoPage(0))
TypeError: connect() slot argument should be a callable or a signal, not 'NoneType'
```

Quick googling leads me to [this post on Stack Overflow](http://stackoverflow.com/questions/15080731/call-a-function-when-a-button-is-pressed-pyqt) and the given answer quoted [this post on someone's blog](http://eli.thegreenplace.net/2011/04/25/passing-extra-arguments-to-pyqt-slot/) with relevant example that uses buttons. From here, I found that there are two methods to reuse the code: `lambda` and `functools.partial`.

Basically, `lambda` is the easy method, often used and friendly to beginners;  `functools.partial` is the reliable method and is more Python-ish. For more information, refer to the following resources.
- [functools in Python 2](https://docs.python.org/2/library/functools.html#functools.partial) on Python Documentation
- [functools in Python 3](https://docs.python.org/3.5/library/functools.html#functools.partial)  on Python Documentation
- [Why is functools.partial necessary](http://stackoverflow.com/questions/3252228/python-why-is-functools-partial-necessary)  on Stack Overflow

**Comparing implementation**

Implementing push buttons using `lambda`:

```
self.pushButton_1.clicked.connect(lambda: self.GotoPage(0))
self.pushButton_2.clicked.connect(lambda: self.GotoPage(1))
self.pushButton_3.clicked.connect(lambda: self.GotoPage(2))
```

Implementing push buttons using `functools`:

```
from functools import partial
...
self.pushButton_1.clicked.connect(partial(self.GotoPage, 0))
self.pushButton_2.clicked.connect(partial(self.GotoPage, 1))
self.pushButton_3.clicked.connect(partial(self.GotoPage, 2))
```

Both `lambda` and `functools` uses same function:

```
def GotoPage(self, num):
    self.stackedWidget.setCurrentIndex(num)
```

**Verdict**

In above case, the difference is just the additional line of `from functools import partial` and the syntax for calling the same function. Both methods are supported in Python 2 and Python 3.

If there is anything else to know, there is one instance that `lambda` was about to disappear in Python 3 but later remained unchanged, according to [AlternateLambdaSyntax](https://wiki.python.org/moin/AlternateLambdaSyntax) on Python Wiki.

Personally, I am tempted to use `lambda` because it is very simple, just add `lambda:` to each slot and I am done. However, thinking of the coding style to be adopted by future contributors, I would prefer `functools` method and encourage more Python-ish coding style instead.

**Comment by ghost** [comment]
**Early preview of 'About' tab**

![gui-420](https://cloud.githubusercontent.com/assets/1223304/20593528/bf0552ae-b26d-11e6-9121-cba06086f166.png)

**Comment by ghost** [comment]
**Updated preview of 'About' tab**

1. Vertical arrangement (GTK+ 3 style)

    ![demo-ui-vl](https://cloud.githubusercontent.com/assets/1223304/22141911/1e59154a-df2f-11e6-8d39-4ef3e5c0bee6.png)

2. Horizontal arrangement (Qt style)

    ![demo-ui-hl](https://cloud.githubusercontent.com/assets/1223304/22141917/23ffff18-df2f-11e6-81d1-e3e3d9eb2e61.png)

**Button behaviour**

Notice that buttons have been improved, which mimic the behaviour of GTK+ 3 widget.

When each button is clicked, button will be highlighted and toggled to indicate currently viewing page of stacked widget. However, unlike GTK+ 3 style, clicking again on the same button will do nothing (this is the simplified behaviour that is easy to implement).

![demo-ui-hl-key](https://cloud.githubusercontent.com/assets/1223304/22142866/e09ad276-df32-11e6-9599-ab47c4c56db0.png)

When each button is focused using keyboard arrow keys, the toggle state will not change (button remained pressed) but highlight colour will be moved to button that has focus. This detail will not be visible while using mouse clicks.

**Widget arrangement**

Horizontal arrangement (Qt style) is preferred for two reasons.

One, horizontal arrangement has more balanced distribution of widget elements compared to vertical counterpart. Applications that follow this arrangement are Qt tools (non-resizable dialogs), VLC and Firefox (resizable window, although the latter is not written using Qt toolkit).

Two, horizontal arrangement has plenty of room of vertical space, which is better than vertical arrangement that barely has any vertical space at optimal window height. This way, logo image size will *not* affect the height of window, which is preferred.

The line spacing between text is balanced, because of these two reasons.

**Comment by ghost** [comment]
Forget above plan, since `gui.py.in` looks messy and I have no more energy and time to look into this matter. The workaround is to edit `gui.ui` in Qt Designer only (so no need to edit the former file for dealing with signals and slots).

Just restore 'About' tab with plainTextEdit and disable text editing feature, then include simple text to note details found in text files installed at following location.

    /usr/share/customizer/contributors
    /usr/share/customizer/copyright

Tested working in local build with screenshot as proof.

![414-fix-gui-about](https://cloud.githubusercontent.com/assets/1223304/22402833/350215f0-e63f-11e6-98ad-771a9b59daa5.png)

This will be included in `finalfix` branch before merging into "old stable".

**Comment by ghost** [comment]
This issue is considered resolved in last update for "old stable".
